hal/src/nRF52840/exflash_hal.c

```diff
@@ -574,3 +584,32 @@ int hal_exflash_special_command(hal_exflash_special_sector_t sp, hal_exflash_com
     hal_exflash_unlock();
     return ret;
 }
+
+int hal_exflash_sleep(bool sleep, void* reserved) {
+    hal_exflash_lock();
+    if (sleep) {
+        // Suspend I2C
+        if (!qspi_initialized) {
+            hal_exflash_unlock();
+            return SYSTEM_ERROR_NONE;
+        }
+        if (qspi_suspended) {
+            hal_exflash_unlock();
+            return SYSTEM_ERROR_NONE;
+        }
+        // Put external flash into sleep and disable QSPI peripheral
+        hal_exflash_special_command(HAL_EXFLASH_SPECIAL_SECTOR_NONE, HAL_EXFLASH_COMMAND_SLEEP, NULL, NULL, 0);
+        hal_exflash_uninit();
+        qspi_suspended = true;
+    } else {
+        // Restore I2C
+        if (!qspi_suspended) {
+            hal_exflash_unlock();
+            return SYSTEM_ERROR_NONE;
+        }
+        hal_exflash_init();
+        qspi_suspended = false;
```

Comment:
This seems to be already done in `hal_exflash_init()`. The errors should also be propagated from `hal_exflash_init()`. It may very well potentially fail, which can as usual lead to filesystem corruption.

Reply:
Fixed, also check other qspi API inside the sleep API